Alerting: Paginate result previews

gillesdemey opened the pull request:
**What is this feature?**

Adds a pagination to the bottom of expression results when previewing alerts from a data query with more than 20 series (multi-dimensional alert rule).

<img width="988" alt="image" src="https://user-images.githubusercontent.com/868844/227185855-24a3afc6-b01f-461d-b5cc-457de557c78c.png">

**Why do we need this feature?**

Large query results could crash the browser (10k+)

**Which issue(s) does this PR fix?**:

Fixes #64830

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.
- [ ] There are no known compatibility issues with older supported versions of Grafana, or plugins.
- [ ] It passes the [Hosted Grafana feature readiness review](https://docs.google.com/document/d/1QL9Ly8KnXzpb6ISbg49pTODRO5mhA5tkkfIZVX6pqQU/edit) for observability, scalability, performance, and security.

soniaAguilarPeiron committed:
Slice to 20 instance rows and render a 'show X more' text

gillesdemey committed:
implement pagination and test
Co-authored-by: 33540275+soniaAguilarPeiron@users.noreply.github.com

konrad147 commented on public/app/features/alerting/unified/components/expressions/Expression.tsx:
```diff
@@ -131,30 +132,82 @@ interface ExpressionResultProps {
   // sometimes we receive results where every value is just "null" when noData occurs
   const emptyResults = isEmptySeries(series);
   const isTimeSeriesResults = !emptyResults && isTimeSeriesFrames(series);
+  const resultPages = chunk(series, PAGE_SIZE);
```
You can try to use `usePagination` hook (from (`public/app/features/alerting/unified/hooks/usePagination.ts`) to remove some of the pagination handling code.

soniaAguilarPeiron replied:
We thought that for this particular case it would be more simple and clear to use `chunk` from `lodash` (we only go forward and backward, we don't go directly to the page)

konrad147 replied:
Hmm, but you had to implement pagination logic once again (pageStart, pageEnd, clamp, chunk)

gillesdemey replied:
Yea I think we can re-use that one, I forgot we had that :D It doesn't seem to expose the `start` and `end` of the current page but that's easy to add to the hook 👍

konrad147 replied:
It exposes `numberOfPages` which should be enough :)

soniaAguilarPeiron replied:
ahh ok! nice!

konrad147 reviewed (approved):
Nice! one nit

gillesdemey committed:
use usePagination

Co-authored-by: konrad147 <konradlalik@gmail.com>

gillesdemey committed:
remove useless comments

gillesdemey commented:
Updated PR with the following changes:

Using `usePagination()` for the preview results.

Made the following changes to the `usePagination()` hook:

* Use clamping in the hook to prevent out-of-bounds access
* Added `nextPage` and `previousPage` which just call `onPageChange()`
* `pageStart` and `pageEnd` exposed
